src/plugins/GLesmos/shaders.ts

```diff
@@ -0,0 +1,419 @@
+      
+      float bestDistance = Infinity;
+      vec4  bestLine     = JFA_undefined;
+      
+      for (int n = 0; n < 9; n++) {
+          
+        vec2 sampleCoord = fragCoord + JFA_kernel[n] / iResolution * stepwidth;
+        vec4 seed        = getPixel( sampleCoord, iChannel0 );
+
+        if( seed == JFA_undefined ) continue; // don't try to use this one
+        float dist = LineSDF( seed * vec4(warp,warp), fragCoord * warp );
+        
+        if (dist < bestDistance){
+          bestDistance = dist;
+          bestLine     = seed;
+        }
+              
+      }
+      
+      return bestLine;
+    }
+
+    void main(){
+
+      vec4 JFA_undefined = vec4(-Infinity);
+
+      vec2 fragCoord = texCoord;
+      
+      if( iInitFlag == 1 ) {  // JFA initialization
+        
+        bool mask = detectSignChange( fragCoord ); // works correctly
+        
+        if( mask ){
+
+          fragCoord = quadTreeSolve(fragCoord, 1.0);
+          fragCoord = quadTreeSolve(fragCoord, 0.5);
+          fragCoord = quadTreeSolve(fragCoord, 0.25);
+          
+          vec2 d = d_f0(fragCoord);
+          d = normalize( vec2(-d.y, d.x) ) / iResolution;
+          
+          outColor = lineToPixel(-d, d, fragCoord);
+        }
+        else {
+          outColor = JFA_undefined;
+        }
+          
+      }
+      else {  // JFA stepping
+        outColor = Step( fragCoord );
+      }
+          
+    }
+
+    //============== END Shadertoy Buffer A ==============//
+  `;
+
+  const shader = getShaderProgram(gl, id, VERTEX_SHADER, source);
+
+  return shader;
+}
+
+export function glesmosGetFinalPassShader(
+  gl: WebGL2RenderingContext,
+  id: string,
+  chunk: GLesmosShaderChunk
+): GLesmosProgram {
+  const source = `${GLESMOS_ENVIRONMENT}
+
+    uniform sampler2D iChannel0;   // storage
+    uniform sampler2D iChannel1;   // cache
+    uniform vec2      iResolution; // canvas size
+    uniform int       iDoOutlines;
+
+    ${GLESMOS_SHARED}
+
+    void main(){
+
+      // fill
+      vec4 test = getPixel( texCoord, iChannel1 );
+      if( test.x > 0.0 ){
+        outColor = mixColor(outColor, ${chunk.color});
+      }
+
+      // lines
+      if( iDoOutlines != 1 ) return;
+      vec4 JFA_undefined = vec4(-Infinity);
+      vec2 warp = iResolution / max(iResolution.x, iResolution.y);
+
+      vec4 seed = getPixel( texCoord, iChannel0 );
+
+      if( seed == JFA_undefined ){ return; }
+
+      float dist = LineSDF( seed * vec4(warp,warp), texCoord * warp ) * max(iResolution.x, iResolution.y);
+
+      float alpha = smoothstep(0.0, 1.0, clamp( dist - float(${chunk.line_width}) * 0.5 + 0.5, 0.0, 1.0 ));
+      outColor = mixColor(outColor, ${chunk.line_color} * vec4(1.0,1.0,1.0,1.0 - alpha));
+    }
+  `;
+
+  const shader = getShaderProgram(gl, id, VERTEX_SHADER, source);
+  gl.useProgram(shader);
+  setUniform(gl, shader, "iDoOutlines", "1i", chunk.line_width > 0 ? 1 : 0);
```

Comment:
Why is `iDoOutlines` a uniform instead of a constant? Probably doesn't really affect performance, but could affect shader compilation time.

Reply:
This uniform lets us avoid compiling a separate more performant shader (or copy-pasting lots of code) for when outlines are off.